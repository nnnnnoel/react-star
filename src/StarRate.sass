.pointer
    cursor: pointer
.left
    margin-left: 5px
.more-left
    margin-left: 10px
.container
    display: inline-block
.app
    display: flex
    align-items: center