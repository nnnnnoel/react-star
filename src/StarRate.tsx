import * as React from "react";
import { ReactComponent as FullStar } from "./star-full.svg";
import { ReactComponent as HalfStar } from "./star-half.svg";
import { ReactComponent as EmptyStar } from "./star-empty.svg";
import { ReactComponent as Reset } from "./reset.svg";
import "./StarRate.sass";

interface Props {
  onChange: (rate: number) => void;
  value: number;
}

const getRate = (x: number): number => Math.min(5, Math.ceil(x / 18.5) / 2);
const toShowable = (x: number): number => Math.min(5, Math.round(x * 2) / 2);

const StarRate: React.FC<Props> = ({ onChange, value }) => {
  const [showableValue, setShowableValue] = React.useState<number>(value);

  // 점수 초기화
  const reset = React.useCallback(() => {
    onChange(0);
  }, [onChange]);

  // 별 onClick 이벤트
  const onClick: React.MouseEventHandler<SVGSVGElement> = React.useCallback(
    e => {
      onChange(getRate(e.pageX));
    },
    [onChange]
  );

  // 별에 마우스 올렸을 때 이벤트
  const onMouseMove: React.MouseEventHandler<SVGSVGElement> = React.useCallback(
    e => {
      if (getRate(e.pageX) !== showableValue)
        setShowableValue(getRate(e.pageX));
    },
    [showableValue]
  );

  // 별에서 마우스 떼었을 때 value로 초기화
  const onMouseLeave: React.MouseEventHandler<SVGSVGElement> = React.useCallback(() => {
    setShowableValue(value);
  }, [value]);

  // value가 바뀔 때마다 보이는 값 조정
  React.useEffect(() => {
    setShowableValue(toShowable(value));
  }, [value]);

  return (
    <div className="app">
      <div {...{ onMouseMove, onMouseLeave }} className="container">
        {Array.from({ length: Math.floor(showableValue) }, (_, i) => {
          return (
            <FullStar
              key={"FullStar" + (i + 1)}
              {...{ onClick }}
              className="pointer left"
              width={32}
              height={32}
            />
          );
        })}
        {showableValue % 1 === 0.5 && (
          <HalfStar
            {...{ onClick }}
            key={"HalfStar" + Math.floor(showableValue + 1)}
            className="pointer left"
            width={32}
            height={32}
          />
        )}
        {showableValue < 5 &&
          Array.from({ length: Math.floor(5 - showableValue) }, (_, i) => {
            return (
              <EmptyStar
                key={"EmptyStar" + Math.ceil(showableValue + i + 1)}
                {...{ onClick }}
                className="pointer left"
                width={32}
                height={32}
              />
            );
          })}
      </div>
      <Reset
        onClick={reset}
        className="pointer more-left"
        width={16}
        height={16}
      />
    </div>
  );
};

export default StarRate;
