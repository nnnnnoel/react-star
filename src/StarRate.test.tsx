import * as React from "react";
import StarRate from "./StarRate";
import { configure, mount } from "enzyme";
import ReactSixteenAdapter from "enzyme-adapter-react-16";

configure({
  adapter: new ReactSixteenAdapter()
});

describe("StarRate run correctly", () => {
  const func = jest.fn();
  const elem = mount(<StarRate onChange={func} value={3.5} />);
  it("renders successfully", () => {
    expect(React.isValidElement(elem.getElement())).toBeTruthy();
    expect(elem).toMatchSnapshot();
    let child = elem.render().children()[0].children;
    expect(child[0].children[0].data).toBe("star-full.svg");
    expect(child[1].children[0].data).toBe("star-full.svg");
    expect(child[2].children[0].data).toBe("star-full.svg");
    expect(child[3].children[0].data).toBe("star-half.svg");
    expect(child[4].children[0].data).toBe("star-empty.svg");
    elem.setProps({
      onChange: func,
      value: 2.5
    });
    child = elem.render().children()[0].children;
    expect(child[0].children[0].data).toBe("star-full.svg");
    expect(child[1].children[0].data).toBe("star-full.svg");
    expect(child[2].children[0].data).toBe("star-half.svg");
    expect(child[3].children[0].data).toBe("star-empty.svg");
    expect(child[4].children[0].data).toBe("star-empty.svg");
    elem.setProps({
      onChange: func,
      value: 3.8
    });
    child = elem.render().children()[0].children;
    expect(child[0].children[0].data).toBe("star-full.svg");
    expect(child[1].children[0].data).toBe("star-full.svg");
    expect(child[2].children[0].data).toBe("star-full.svg");
    expect(child[3].children[0].data).toBe("star-full.svg");
    expect(child[4].children[0].data).toBe("star-empty.svg");
    elem.setProps({
      onChange: func,
      value: 5.7
    });
    child = elem.render().children()[0].children;
    expect(child[0].children[0].data).toBe("star-full.svg");
    expect(child[1].children[0].data).toBe("star-full.svg");
    expect(child[2].children[0].data).toBe("star-full.svg");
    expect(child[3].children[0].data).toBe("star-full.svg");
    expect(child[4].children[0].data).toBe("star-full.svg");
    expect(child[5]).toBeUndefined();
    elem.setProps({
      onChange: func,
      value: 0
    });
    child = elem.render().children()[0].children;
    expect(child[0].children[0].data).toBe("star-empty.svg");
    expect(child[1].children[0].data).toBe("star-empty.svg");
    expect(child[2].children[0].data).toBe("star-empty.svg");
    expect(child[3].children[0].data).toBe("star-empty.svg");
    expect(child[4].children[0].data).toBe("star-empty.svg");
    elem.setProps({
      onChange: func,
      value: 4.3
    });
    child = elem.render().children()[0].children;
    expect(child[0].children[0].data).toBe("star-full.svg");
    expect(child[1].children[0].data).toBe("star-full.svg");
    expect(child[2].children[0].data).toBe("star-full.svg");
    expect(child[3].children[0].data).toBe("star-full.svg");
    expect(child[4].children[0].data).toBe("star-half.svg");
    elem.setProps({
      onChange: func,
      value: 4.2
    });
    child = elem.render().children()[0].children;
    expect(child[0].children[0].data).toBe("star-full.svg");
    expect(child[1].children[0].data).toBe("star-full.svg");
    expect(child[2].children[0].data).toBe("star-full.svg");
    expect(child[3].children[0].data).toBe("star-full.svg");
    expect(child[4].children[0].data).toBe("star-empty.svg");
  });
});
