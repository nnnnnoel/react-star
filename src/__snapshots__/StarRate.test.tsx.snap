// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarRate run correctly renders successfully 1`] = `
<StarRate
  onChange={[MockFunction]}
  value={3.5}
>
  <div
    className="app"
  >
    <div
      className="container"
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
    >
      <ForwardRef(SvgStarFull)
        className="pointer left"
        height={32}
        key="FullStar1"
        onClick={[Function]}
        width={32}
      >
        <svg
          className="pointer left"
          height={32}
          onClick={[Function]}
          width={32}
        >
          star-full.svg
        </svg>
      </ForwardRef(SvgStarFull)>
      <ForwardRef(SvgStarFull)
        className="pointer left"
        height={32}
        key="FullStar2"
        onClick={[Function]}
        width={32}
      >
        <svg
          className="pointer left"
          height={32}
          onClick={[Function]}
          width={32}
        >
          star-full.svg
        </svg>
      </ForwardRef(SvgStarFull)>
      <ForwardRef(SvgStarFull)
        className="pointer left"
        height={32}
        key="FullStar3"
        onClick={[Function]}
        width={32}
      >
        <svg
          className="pointer left"
          height={32}
          onClick={[Function]}
          width={32}
        >
          star-full.svg
        </svg>
      </ForwardRef(SvgStarFull)>
      <ForwardRef(SvgStarHalf)
        className="pointer left"
        height={32}
        key="HalfStar4"
        onClick={[Function]}
        width={32}
      >
        <svg
          className="pointer left"
          height={32}
          onClick={[Function]}
          width={32}
        >
          star-half.svg
        </svg>
      </ForwardRef(SvgStarHalf)>
      <ForwardRef(SvgStarEmpty)
        className="pointer left"
        height={32}
        key="EmptyStar5"
        onClick={[Function]}
        width={32}
      >
        <svg
          className="pointer left"
          height={32}
          onClick={[Function]}
          width={32}
        >
          star-empty.svg
        </svg>
      </ForwardRef(SvgStarEmpty)>
    </div>
    <ForwardRef(SvgReset)
      className="pointer more-left"
      height={16}
      onClick={[Function]}
      width={16}
    >
      <svg
        className="pointer more-left"
        height={16}
        onClick={[Function]}
        width={16}
      >
        reset.svg
      </svg>
    </ForwardRef(SvgReset)>
  </div>
</StarRate>
`;
